python/ray/dag/compiled_dag_node.py

```diff
@@ -1564,42 +1570,48 @@ def _execute_until(
 
         TODO(rui): catch the case that user holds onto the CompiledDAGRefs
         """
-        from ray.dag import DAGContext
+        if self._max_execution_index < execution_index:
+            from ray.dag import DAGContext
 
-        ctx = DAGContext.get_current()
-        if timeout is None:
-            timeout = ctx.retrieval_timeout
+            ctx = DAGContext.get_current()
+            if timeout is None:
+                timeout = ctx.retrieval_timeout
 
-        while self._max_execution_index < execution_index:
-            if self._max_execution_index + 1 == execution_index:
-                # Directly fetch and return without buffering
+            while self._max_execution_index < execution_index:
+                if len(self._result_buffer) >= self._max_buffered_results:
```

Comment:
I think this is wrong now. We should do len(self._result_buffer) * num_output_channels

Reply:
can you also add a unit test? (or modify existing test)

Reply:
Agree. `len(self._result_buffer)` indicates the number of DAG executions while `len(self._result_buffer) * num_output_channels` represents the total number of outputs. I will adjust the naming of `_max_buffered_results` to `_max_buffered_executions` (any other suggestions?) and add a unit test.

Reply:
Actually, I think we need to change `max_buffered_results` to `max_buffered_executions` if we were to change `_max_buffered_results` to `_max_buffered_executions`. Since there are quite a few references to `max_buffered_results` in the repo (including developer APIs and `python/ray/dag/context.py` which doesn't have much relevance to the changes in this PR), I'd suggest to not modify the naming now. We can track it as a bug and address it in a follow-up.

As for the unit test, I'll add one tomorrow.

Reply:
Done adding a unit test for cases when there are multiple outputs. Now, if not all results from an execution index are fetched with `get()`, that execution index will still count towards the number of buffered results. Let me know if this sounds okay to you!